### ISSUE TITLE
publish still "doesn't publish" (without -r)

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#1426]
e.g. I am in the top of our crawled mega-dataset, and trying to publish it after I have added workshops/ subdataset:

```shell
(git)smaug:/mnt/btrfs/datasets/datalad/crawl[master]git
$> datalad publish --to=datalad-public
[INFO   ] Publishing <Dataset path=/mnt/btrfs/datasets/datalad/crawl> to datalad-public
yoh@datasets.datalad.org's password:
Permission denied, please try again.
yoh@datasets.datalad.org's password:
No handlers could be found for logger "git.remote"
1 item was published:
Dataset: /mnt/btrfs/datasets/datalad/crawl


$> git lg
* 4ea5f63 - (HEAD -> master) [DATALAD] aggregated meta data (22 minutes ago) [Yaroslav Halchenko]
* c544eac - aggregated meta-data for the workshops which now includes nipype 2017 (23 minutes ago) [Yaroslav Halchenko]
* 2699b2f - [DATALAD] added content (31 minutes ago) [Yaroslav Halchenko]
* 9a87f45 - (datalad-public/master) [DATALAD] aggregated meta data (4 days ago) [Yaroslav Halchenko]
* d581162 - [DATALAD] auto-saved changes (4 days ago) [Yaroslav Halchenko]
```

so as you can see datalad-public/master is still  at older point

NB1 adding -r makes action to take long time, so I guess it does its "lets get entire hierarchy just in case" although there were no changes to other datasets but 'workshops' since the last push.
But it seems to do the right thing
```shell
$> datalad publish --to=datalad-public -r
[INFO   ] Publishing <Dataset path=/mnt/btrfs/datasets/datalad/crawl/workshops> to datalad-public
yoh@datasets.datalad.org's password:
Permission denied, please try again.
yoh@datasets.datalad.org's password:
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/workshops/nipype-2017/ds000114> ...
[INFO   ] Publishing <Dataset path=/mnt/btrfs/datasets/datalad/crawl> to datalad-public
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/Philips-20131217> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/bids_test4-20161014/phantom-8> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/bids_test3> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/PHANTOM-20160909> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/labs/haxby/raiders> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/bids_test4-20161014/phantom-9> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/bids_test4-20161014/phantom-4> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/bids_test4-20161014/phantom-5> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/bids_test4-20161014/phantom-6> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/bids_test4-20161014/phantom-7> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/bids_test4-20161014/phantom-1> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/bids_test4-20161014/phantom-2> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/bids_test4-20161014/phantom-3> ...
[INFO   ] Publishing data of dataset <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dicoms/dartmouth-phantoms/PHANTOM1_3> ...
2 items were published:
Dataset: /mnt/btrfs/datasets/datalad/crawl/workshops
Dataset: /mnt/btrfs/datasets/datalad/crawl

datalad publish --to=datalad-public -r  96.94s user 40.45s system 33% cpu 6:47.09 total
```

and then just keeps saying that it is publishing it although it was already fully up-to-date
```shell
$> datalad publish --to=datalad-public -r
[INFO   ] Publishing <Dataset path=/mnt/btrfs/datasets/datalad/crawl> to datalad-public 
[INFO   ] [up to date]
|  
1 item was published:
Dataset: /mnt/btrfs/datasets/datalad/crawl
```

NB2 it would have been nice if it announced first how many datasets it is going to publish or any other additional detail relevant if it knows ;)

**Comment by mih** [comment]
NB 1: should be a non-issue in #1544 now.

**Comment by mih** [comment]
This is fixed in #1544 (dedicated test included).

**Comment by mih** [comment]
this should be good now
